provider/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,429 @@
+				if _, ok := conf.Services["default-backend"]; ok {
+					log.FromContext(ctx).Error("The default backend already exists.")
+					continue
+				}
+
+				service, err := loadService(client, ingress.Namespace, *ingress.Spec.Backend)
+				if err != nil {
+					log.FromContext(ctx).
+						WithField("serviceName", ingress.Spec.Backend.ServiceName).
+						WithField("servicePort", ingress.Spec.Backend.ServicePort.String()).
+						Errorf("Cannot create service: %v", err)
+					continue
+				}
+
+				conf.Routers["/"] = &config.Router{
+					Rule:     "PathPrefix(`/`)",
+					Priority: math.MinInt32,
+					Service:  "default-backend",
+				}
+
+				conf.Services["default-backend"] = service
+			}
+		}
+		for _, rule := range ingress.Spec.Rules {
+			if err := checkStringQuoteValidity(rule.Host); err != nil {
+				log.FromContext(ctx).Errorf("Invalid syntax for host: %s", rule.Host)
+				continue
+			}
+
+			for _, p := range rule.HTTP.Paths {
+				service, err := loadService(client, ingress.Namespace, p.Backend)
+				if err != nil {
+					log.FromContext(ctx).
+						WithField("serviceName", p.Backend.ServiceName).
+						WithField("servicePort", p.Backend.ServicePort.String()).
+						Errorf("Cannot create service: %v", err)
+					continue
+				}
+
+				if err = checkStringQuoteValidity(p.Path); err != nil {
+					log.FromContext(ctx).Errorf("Invalid syntax for path: %s", p.Path)
+					continue
+				}
+
+				serviceName := ingress.Namespace + "/" + p.Backend.ServiceName + "/" + p.Backend.ServicePort.String()
+
+				var rules []string
+				if len(rule.Host) > 0 {
+					rules = []string{"Host(`" + rule.Host + "`)"}
+				}
+
+				if len(p.Path) > 0 {
+					rules = append(rules, "PathPrefix(`"+p.Path+"`)")
+				}
+
+				conf.Routers[strings.Replace(rule.Host, ".", "-", -1)+p.Path] = &config.Router{
+					Rule:    strings.Join(rules, " && "),
+					Service: serviceName,
+				}
+
+				conf.Services[serviceName] = service
+			}
+		}
+	}
+
+	conf.TLS = getTLSConfig(tlsConfigs)
+	return conf
+}
+
+func shouldProcessIngress(ingressClass string, ingressClassAnnotation string) bool {
+	return ingressClass == ingressClassAnnotation ||
+		(len(ingressClass) == 0 && ingressClassAnnotation == traefikDefaultIngressClass)
+}
+
+func getTLS(ctx context.Context, ingress *v1beta1.Ingress, k8sClient Client, tlsConfigs map[string]*tls.Configuration) error {
+	for _, t := range ingress.Spec.TLS {
+		if t.SecretName == "" {
+			log.FromContext(ctx).Debugf("Skipping TLS sub-section: No secret name provided")
+			continue
+		}
+
+		configKey := ingress.Namespace + "/" + t.SecretName
+		if tlsConfig, tlsExists := tlsConfigs[configKey]; !tlsExists {
+			secret, exists, err := k8sClient.GetSecret(ingress.Namespace, t.SecretName)
+			if err != nil {
+				return fmt.Errorf("failed to fetch secret %s/%s: %v", ingress.Namespace, t.SecretName, err)
+			}
+			if !exists {
+				return fmt.Errorf("secret %s/%s does not exist", ingress.Namespace, t.SecretName)
```

Comment:
Due to https://github.com/containous/traefik/issues/3667, we may want to use this opportunity to resolve this.